Brim: optionally generate brim from Elephant Foot Compensation outline

kisslorand opened the pull request:
## Summary

This PR adds an optional mode that makes **brim generation follow the Elephant Foot Compensation (EFC)–adjusted first-layer outline**, instead of the original model outline.

When enabled, the brim is generated relative to the **actual first-layer footprint produced after EFC**, keeping the brim aligned with compensated perimeters and restoring its intended role in cases where EFC significantly alters the effective first-layer footprint.

Default behavior is unchanged.

## Motivation / Problem

Elephant Foot Compensation modifies the effective footprint of the first layer.
However, brim generation has historically referenced the raw model outline, which can diverge significantly from the compensated geometry.

This mismatch can lead to two practical issues:
- With **aggressive EFC**, the brim may no longer touch the compensated first layer, reducing its effectiveness for adhesion.
- When EFC is used **intentionally and/or effectively to alter base geometry** (e.g. chamfer-like effects and/or printability constraints), the brim is generated relative to geometry that is no longer representative of the printed part.

In both cases, the brim may fail to fulfill its purpose due to outline mismatch, not user intent.

## What this PR does

This PR introduces a selectable option that:
- Uses the already computed EFC-adjusted first-layer outline as the reference geometry for brim generation
- Aligns brim geometry with the same outline used by compensated perimeters
- Applies consistently to brim ears, auto-brim heuristics, and per-volume-group processing

The EFC-adjusted outline is reused from existing first-layer geometry, avoiding any attempt to approximate compensation via offsets.

## Why this is not a simple offset

Elephant Foot Compensation is not a uniform geometric offset:
- It is flow-aware
- It avoids generating unprintable thin features
- It may intentionally alter the object’s base geometry
- It can differ per island or volume group

For correctness, this PR reuses the **already computed first-layer geometry**, rather than attempting to reconstruct it from the raw model outline.

## Presets and Backward Compatibility

This change is fully backward compatible with existing presets.
- The new behavior is opt-in and disabled by default.
- Existing profiles and saved projects will continue to generate brims using the original model outline unless the option is explicitly enabled.
- No existing preset values are reinterpreted or migrated.
- Enabling this option only affects how the brim reference outline is chosen; it does not modify Elephant Foot Compensation itself, extrusion parameters, or first-layer settings.

As a result, existing presets retain identical slicing output unless the user deliberately enables the new option.

## User-facing warning (documented in tooltip)

- With aggressive or geometry-altering EFC, enabling this option may cause brim lines to overlap higher layers.
- Disabling it in such cases may leave the brim detached from the compensated first layer, reducing its adhesion benefit.

This reflects a real geometric trade-off and is explicitly communicated to the user.

## Examples:

The object to be sliced:
<img width="763" height="651" alt="image" src="https://github.com/user-attachments/assets/500ca720-d52e-4926-9072-c8bcbd165dbf" />

The 2nd layer of the sliced object (reference for first layer geometry deformation):
<img width="721" height="642" alt="image" src="https://github.com/user-attachments/assets/0f02c7d1-cfa2-4603-ab8b-56317dd4f00e" />

The followings are sliced with EFC 0.6mm and brim-object gap 0.2:

Outer brim, `Brim follows compensated outline` __disabled__:
<img width="879" height="785" alt="image" src="https://github.com/user-attachments/assets/34778b9c-6a4d-4232-a2b0-792d79888139" />

Outer brim, `Brim follows compensated outline` __enabled__:
<img width="875" height="787" alt="image" src="https://github.com/user-attachments/assets/bee47a94-24dc-4969-93b6-e6ab86287fdc" />

Painted brim, `Brim follows compensated outline` __disabled__:
<img width="667" height="602" alt="image" src="https://github.com/user-attachments/assets/f6df6c34-aae7-4c10-ba8f-720cbd62e838" />

Painted brim, `Brim follows compensated outline` __enabled__:
<img width="673" height="609" alt="image" src="https://github.com/user-attachments/assets/2d0c2429-97d2-4e78-b2cd-a08f105f7083" />

## Conclusions

The screenshots above show that, in certain cases, Elephant Foot Compensation meaningfully alters the effective first-layer footprint relative to the original model outline.

In such cases, generating the brim from the unmodified geometry can cause it to lose proper contact with the compensated first-layer perimeter, reducing its effectiveness for first-layer adhesion.
Because the brim no longer follows the actual printed footprint, it also becomes unreliable as a visual or mechanical reference when evaluating or calibrating Elephant Foot Compensation.

Using the EFC-adjusted first-layer outline for brim generation makes the compensated first-layer toolpath explicit and consistent, ensuring the brim reflects the same geometry used to generate the compensated perimeters when Elephant Foot Compensation is applied.

**This is not universally required**: when Elephant Foot Compensation is mild, the compensated footprint’s geometry may remain close to the original model outline.
The usefulness of this option becomes apparent when Elephant Foot Compensation significantly alters the effective first-layer footprint, either due to strong compensation values or due to printability constraints that intentionally modify base geometry.

Fixes #7494
Fixes #10877
Fixes #3720

Thank you @RF47!

>Note: Translations were generated using online translation tools; native speakers are welcome to suggest corrections.

kisslorand committed:
Brim can follow EFC outline

kisslorand committed:
Optimization

RF47 commented on localization/i18n/es/OrcaSlicer_es.po:
```diff
@@ -12147,6 +12147,28 @@ msgstr ""
+"such cases may leave the brim detached from the compensated first layer, "
+"reducing its effectiveness for first-layer adhesion."
+msgstr ""
+"Utilice el esquema de compensación de pata de elefante para la generación de Borde de adherencia.\n"
```
"Utilice el contorno de compensación de pata de elefante para la generación de Borde de adherencia.\n"
"El borde de adherencia se genera a partir de la geometría compensada de la primera capa en lugar "
"del contorno del modelo original, lo que garantiza el contacto con los perímetros compensados.\n"
"\n"

RF47 commented on localization/i18n/es/OrcaSlicer_es.po:
```diff
@@ -12147,6 +12147,28 @@ msgstr ""
+"\n"
+"Con una compensación de pata de elefante agresiva o que altera la geometría, habilitar "
+"esta opción puede hacer que Borde de adherencia se fusione con las capas superiores. "
+"Deshabilitarlo en tales casos puede dejar Borde de adherencia desprendido de la primera "
```
"Deshabilitarlo en tales casos puede dejar el borde de adherencia desprendido de la primera "
"capa compensada, lo que reduce su efectividad para la adhesión de la primera capa."

PhilBaz commented:
Nice!

Thank you :)

kisslorand committed:
Update Spanish EFC brim description

Adopt reviewer-proposed wording from RF47.

Co-authored-by: RF47 <RF47@users.noreply.github.com>

igiannakas commented:
Could you do me a favour please and tag any orca specific changes vs. bambu using the comment //ORCA: . This helps a ton when reviewing merge commits from upstream bambu so we don't end up causing regressions when pulling in commits from upstream :)

I've merged this PR on my test branch to use day to day.

igiannakas commented:
In addition to your testing, performed test on multicolour painted object and setting these per object and at the global level.

Works well.
![image](https://github.com/user-attachments/assets/84b7cf2e-7ee4-43fb-8397-932583d302b6)

@SoftFever I think this is a good compromise? We could "hide" it under "dev mode" if needs must?

igiannakas committed:
Merge branch 'main' into Brim-follows-outline

igiannakas committed:
Merge branch 'main' into Brim-follows-outline

kisslorand committed:
Tag Orca specific changes

Tag Orca specific changes vs. Bambu using the comment //ORCA: . This helps when reviewing merge commits from upstream Bambu so we don't end up causing regressions when pulling in commits from upstream

kisslorand commented:
> Could you do me a favour please and tag any orca specific changes vs. bambu using the comment //ORCA: . This helps a ton when reviewing merge commits from upstream bambu so we don't end up causing regressions when pulling in commits from upstream :)

Done. Please feel free to add suggestions if I missed some.

SoftFever reviewed (commented):
@kisslorand thank you!
Overall, this looks good to me.  
I suggested an update for the tooltip content—could you take a look?

kisslorand commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -1562,6 +1562,17 @@ void PrintConfigDef::init_fff_params()
+    def = this->add("brim_use_efc_outline", coBool);
+    def->label = L("Brim follows compensated outline");
+    def->category = L("Support");
+    def->tooltip = L("Use Elephant Foot Compensation outline for brim generation.\n"
```
I understand the intent behind rephrasing the tooltip, but I’m a bit reluctant to redo it unless it’s strictly necessary.

One technical clarification that may help here: **X-Y shrinkage, contour, and hole compensation are already baked into brim generation today.**
The brim is generated from geometry that has already passed through those global X-Y adjustments.

What this toggle adds specifically is **alignment with first-layer–only geometry changes introduced by Elephant Foot Compensation**, which previously were not reflected in the brim reference. It doesn’t newly introduce or broaden X-Y compensation behavior.

Given that:
- the current tooltip already describes this behavior accurately and narrowly, and
- any change to the tooltip text (regardless of size) requires editing and re-translating all language files,

I’d prefer to keep the existing wording unless there’s a concrete inaccuracy or user confusion it needs to address.
If there’s a specific part you consider misleading, I’m happy to adjust just that.

SoftFever replied:
Sounds good! Thank you 👍

kisslorand committed:
Tooltip update

PhilBaz commented:
@kisslorand & @RF47 

Thanks again for your work on this guy's. Very much appreciated!

SoftFever commented:
@RF47 @kisslorand @igiannakas @PhilBaz 
Merged.
Thanks everyone for the amazing work!

fyrak1s commented:
I am on Orca Sliver 2.3.2 and this does not seem to have been implemented for mouse ears? The mouse ears are not attached to the actual object when elephant foot compensation is set. Thanks!

kisslorand commented:
<img width="1351" height="686" alt="image" src="https://github.com/user-attachments/assets/fddc5bec-0226-4b87-bfd1-0f1b9547b156" />
<br>
<br>
Mouse ears seems pretty attached to me...

Did you enable "Brim follows compensated outline"?

<img width="456" height="88" alt="image" src="https://github.com/user-attachments/assets/0a0057ab-f812-46f0-98e9-5f3d9826ce06" />

fyrak1s commented:
@kisslorand no, I've missed that setting.

I have just tested this setting and it works!

I wouldn't say that it is very intuitive to 'link' together (in the users head) elephant compensation and this brim setting (unless they would read the tool tip).
Maybe in the elephant foot setting there should be a tip in its description ([like its wiki entry does at the end](https://www.orcaslicer.com/wiki/print_settings/quality/quality_settings_precision#elephant-foot-compensation)). Or the name of the setting could include the term "elephant foot compensation" like the [wiki entry](https://www.orcaslicer.com/wiki/print_settings/others/others_settings_brim#brim-use-efc-outline).

But still I am very happy we now have this option. Thank you!

EDIT:
In a nutshell, what I'd suggest:
1. Rename this new setting from "_Brim follows compensated outline_" to "_Compensate for Elephant Foot outline_" (which is less generic and the setting applies exclusively for EFC anyways!).
2. To the bottom of the tooltip of the original "_Elephant foot compensation_" setting add:
    > Note: Setting this to an overly high value may cause the brim to end up disconnected in the final print! For such case, consider activating "Compensate for Elephant Foot outline" in brim settings.

That should make it very clear & discoverable without adding complexity.
